### DISCUSSION TITLE
Incorporating RobotRaconteur into a C++ codebase Makefile

### DISCUSSION THREAD

**Discussion by PeterCho** [robotraconteur/robotraconteur#199] [Q&A]
Hello.

I am a RobotRaconteur newbie.  I just learned about RR a few days ago.
I'd like to use it to exchange messages between a main C++ application
that performs 2D/3D computer vision operations and a robotic arm.

A colleague at my workplace who is familiar with RobotRaconteur set up a dummy
python server which plays the role of the robot arm receiving
messages.  He also set up an example C++ client that broadcasts
position and orientation waypoints for the robot arm to follow.

I first followed my colleague's installation instructions:

   **sudo add-apt-repository ppa:robotraconteur/ppa
   sudo apt-get update
   sudo apt-get sudo add-apt-repository ppa:robotraconteur/ppa
   sudo apt-get update
   sudo apt-get install python3-robotraconteur
   sudo apt-get install robotraconteur-dev
   sudo apt-get install librobotraconteur-companion-devinstall python3-robotraconteur
   sudo apt-get install robotraconteur-dev
   sudo apt-get install librobotraconteur-companion-dev**
 
I then used his CMake setup to build and run his dummy RR C++ client
and python server programs.  They communicate fine with each other on
my ubuntu 20.4 machine.

His example C++ client program is reasonably straightfoward for me to
follow.  So I'd really like to integrate it into my own C++ codebase
so that I can use my computer vision algorithms to control a robot arm
we have in our lab.

Unfortunately, I'm running into problems with rebuilding my
colleague's C++ example program inside my own C++ codebase.  My
codebase is based upon a large Makefile (rather than a CMake build
system).  I added paths into my Makefile to the two libRobotRaconteur
libraries that appear to have been installed via the "apt-get"
commands above:

**LIBS =  ...
        -L/usr/lib/ \
        /usr/lib/x86_64-linux-gnu/libRobotRaconteurCore.a \
        /usr/lib/x86_64-linux-gnu/libRobotRaconteurCompanion.a \
        -lboost_regex  -lboost_thread -lboost_filesystem \
        -lboost_random -lboost_chrono -lboost_program_options \
        ...**

I also empirically found that the boost libraries above had to be
included into LIBS before RobotRaconteur could build.

But the linker says it cannot find a definition for
RobotRaconteur::MessageStringPtr::MessageStringPtr:

**/usr/bin/ld: xarm_rr_example.o: in function `RobotRaconteur::RRBaseArray::GetTypeString()':
xarm_rr_example.cc:(.text._ZN14RobotRaconteur11RRBaseArray13GetTypeStringEv[_ZN14RobotRaconteur11RRBaseArray13GetTypeStringEv]+0xc2): undefined reference to `RobotRaconteur::MessageStringPtr::MessageStringPtr(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&)'
/usr/bin/ld: xarm_rr_example.o: in function `boost::detail::sp_if_not_array<analog::xarm_rr::analog__xarm_rrFactory>::type boost::make_shared<analog::xarm_rr::analog__xarm_rrFactory>()':
xarm_rr_example.cc:(.text._ZN5boost11make_sharedIN6analog7xarm_rr22analog__xarm_rrFactoryEJEEENS_6detail15sp_if_not_arrayIT_E4typeEDpOT0_[_ZN5boost11make_sharedIN6analog7xarm_rr22analog__xarm_rrFactoryEJEEENS_6detail15sp_if_not_arrayIT_E4typeEDpOT0_]+0x1e): undefined reference to `vtable for analog::xarm_rr::analog__xarm_rrFactory'
collect2: error: ld returned 1 exit status
make: *** [Makefile:576: xarm_rr_example] Error 1**

Is there another RR-specific library or object file that I need to
include into my Makefile beyond the two RobotRaconteur .a files listed above?

My C++ Makefile also has not yet been told anything about the
analog::xarm_rr:: namespace.  In my colleague's original Cmake version
which did correctly build and run on my ubuntu 20.4 linux
machine, I see object files

   **analog__xarm_rr_stubskel.cpp.o
   xarm_rr_example.cpp.o**

buried deep inside a directory tree.  So I'm guessing that my Makefile
needs probably needs to incorporate something like the
"analog__xarm_rr_stubskel.cpp.o" file.

I'd be very grateful for any advice on how to integrate RobotRaconteur
into my C++ Makefile.

Thanks very much.

Peter

**Comment by johnwason** [comment]
The cmake build system generates some cpp/h files and sets a few compiler definitions that need to be added to your makefile. My suggestion is to use the “Makefile” cmake generator from cmake for your example and then look at the makefiles that are generated. 

The cmake system automatically calls RobotRaconteurGen which is a command line tool to generate code for service types.

**Reply by PeterCho** [reply]
Thanks for the quick response.  My knowledge of cmake, unfortunately, is quite limited.  But so far as I can tell, the cmake system set up by my colleague is calling a Unix Makefile generator.  After poking around in the Makefiles created via cmake, I see they wrote a file deep inside a directory tree called "link.txt".  Its contents appear to confirm that the two .o object files above must indeed be supplied to the linker (with a -rdynamic flag that I need to learn about):

**/usr/bin/c++ -g -rdynamic CMakeFiles/xarm_rr_example.dir/analog__xarm_rr_stubskel.cpp.o CMakeFiles/xarm_rr_example.dir/xarm_rr_example.cpp.o -o xarm_rr_example  /usr/lib/x86_64-linux-gnu/libRobotRaconteurCore.a /usr/lib/x86_64-linux-gnu/libRobotRaconteurCompanion.a /usr/lib/x86_64-linux-gnu/libboost_date_time.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_filesystem.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_system.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_regex.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_chrono.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_atomic.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_thread.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_random.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_program_options.so.1.71.0 /usr/lib/x86_64-linux-gnu/libRobotRaconteurCore.a /usr/lib/x86_64-linux-gnu/libboost_date_time.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_filesystem.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_regex.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_chrono.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_thread.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_atomic.so.1.71.0 -lpthread /usr/lib/x86_64-linux-gnu/libboost_random.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_system.so.1.71.0 /usr/lib/x86_64-linux-gnu/libboost_program_options.so.1.71.0 /usr/lib/x86_64-linux-gnu/libssl.so /usr/lib/x86_64-linux-gnu/libcrypto.so -lpthread -lrt -lz -ldl** 

link.txt also contains the two libRobotRaconteur .a files and several boost libraries which makes sense.  So after doing more googling, I'll try explicitly adding the two .o files into my Makefile with the hope that will allow the linker to find what it needs for building the RobotRaconteur example file inside my C++ codebase.

**Reply by johnwason** [reply]
I wrote a quick page describing how to use Makefiles.

https://github.com/robotraconteur/robotraconteur/wiki/Makefile-Compilation

**Comment by johnwason** [comment]
That is the linker step. There are other entries that compile \*.cpp to \*.cpp.o.  Look for -DROBOTRACONTEUR_\* flags.

**Comment by PeterCho** [comment]
Thanks!  I've been working with Honglu He over the past few weeks trying to
get a robotic Xarm integrated into a vision + language system that I've
been developing with other colleagues inside Analog Garage.  (Our lab is
located in downtown Boston.)  So far, I'm just sending tcp-ip string
messages from a primary app to Honglu's python scripts which drive the
arm.  But before Honglu finishes his internship at the end of April, I'd
really like to replace the current tcp-ip messaging hack with Robot
Raconteur.

If you're curious, I can let you know how our experiments with Honglu
proceed in the coming days...

Peter

**Reply by johnwason** [reply]
Sure, very curious about your work!

**Comment by PeterCho** [comment]
Hi John.

Tonight I was able to morph your makefile_example_service/client into my
ancient/monolithic Makefile that I use for all of my C++ program
development.  So I'm now able to build your simple server/client example
outside of Cmake which is really helpful to me.  I'll next try to get a
variant of your example that Honglu wrote a few weeks ago to also build in
my C++ makefile environment.

If that's successful,  Honglu and I can try replacing our current
working-but-hacky tcp/ip string message passing between my C++ vision +
language perception app and his Xarm python script with robot raconteur
calls.  It'd be great if we can get raconteur communications working before
Honglu finishes his Analog Garage internship at the end of this month.

Peter

PS:  A few days ago, Honglu and I filmed a 4 minute iphone movie of his
operating a robotic arm pick & place demonstration via voice commands.  I
can send you a copy of that movie which you might get a kick out of
seeing.  But I'd feel better emailing a video link to some personal account
for you rather than to this public github address...

**Reply by PeterCho** [reply]
Over the past month, I've been able to get RR to build and run on my ubuntu 20.4 linux box.  On that machine, I'm working with a Makefile rather than cmake setup.  After I followed examples from you and Honglu, it wasn't too hard to get my own little server/client app to communicate via RR.

Now I'm trying to get that same server/client pair to build on an ubuntu 22.4 laptop.  And this time, I'm building using cmake rather than from a Makefile.  I've been getting lots of help from one of Analog Garage's best software engineers.  Yet there remains a serious problem which we don't (yet) know how to resolve.  

Under ubuntu 22.4, there appears to be a basic problem with boost when building an app that needs RR.  In particular, it looks like the version of boost on my 22.4 laptop uses the symbol "I" all over the place.  Yet this seems to be clashing with I being defined in other libraries as sqrt(-1).  

Are you aware of any problems with boost, sqrt(-1) or "I" when it comes to building and running RR within ubuntu 22.4 setups?

**Reply by PeterCho** [reply]
![eigen_error](https://github.com/robotraconteur/robotraconteur/assets/3648253/3980bf76-40c8-4afe-8a9b-5c50e5014902)

**Reply by johnwason** [reply]
Yikes, that looks like a very painful symbol clash. This is not a problem in RR itself, since the library doesn't define an `I` symbol, but rather something with Boost and Eigen. My guess is that there is a `using namespace Eigen` or `using namespace boost` somewhere that is causing this. The `using namespace` directives will import all symbols into the global namespace and can cause this type of clash. This is especially bad in header files since it can pollute otherwise working code. In general, it is best to avoid `using namespace` unless you are very confident in what is going on with various symbols. For Robot Raconteur, I recommend using aliases:

```
namespace RR=RobotRaconteur;
namespace RRC=RobotRaconteur::Companion;
```

etc, and only in cpp files. Never use namespace imports or aliases in the global namespace in header files.

Sometimes changing the order of include files can also help.

I hope this helps, but this type of issue is one of the many pain points in C++ and one of the reasons I am looking forward to something like Rust supplanting it.

**Reply by johnwason** [reply]
So I was not correct, the `I` symbol is coming from `complex.h` in the GCC standard library, not Eigen. My suggestion is to do the following at the top of the CPP file before any other includes:

```
#include <complex.h>
#undef I

#include <...>
```

That should include the complex.h file and undefine I. Because header files are only included once, this should prevent I from being redefined.

**Reply by johnwason** [reply]
You will need to place this at the top of all CPP files with the problem, before any other includes.

**Comment by PeterCho** [comment]
Thanks for this suggestion!  Following it, my C++ main program now looks
like

#include <iostream>
#include <string>
#include <vector>

#include <complex.h>
#undef I

#include <Eigen/Dense>
#include <RobotRaconteur.h>
#include "RobotRaconteurCompanion/Converters/EigenConverters.h"

blah blah

And this main program builds.  Tomorrow I'll test this C++ server program
to see if it can send time stamps, 3D positions and 4D quaternions via RR
to Honglu's python client which drives our robot arm.  I hope so...


Peter

**Comment by PeterCho** [comment]
Over the past few days as I've been trying to install RR on various linux boxes, I found that the following apt-get command was missing from the directions:

sudo apt-get install librobotraconteur-companion-dev

At least I don't remember seeing this mentioned anywhere...

**Reply by johnwason** [reply]
Thanks, I am working on improving the overall documentation.

**Comment by PeterCho** [comment]
I'm writing to ask two more questions related to working with RR in C++

1.  I have a header file camera.h for a camera class which contains guards:

_#ifndef CAMERA_H
#define CAMERA_H

#include <iostream>
#include <string>
#include <vector>
...
#endif // camera.h_

This particular class contains a member function

_void get_b_coeffs(
      double& B0, double& Bu, double& Bv, ...)_



Today I observed some variable name clash between B0 in the definition
above and something inside of RR.  I ended up having to rename B0 -->
Bzero to circumvent the name clash.  This conflict surprises me since
I thought my CAMERA_H guard would prevent such a variable clash.  Any ideas?

**Reply by johnwason** [reply]
That should not be an issue with RR since the variables are in a namespace. This appears to be coming from another header file. Unfortunately there is not much I can do about poorly designed header files from other projects.

**Comment by PeterCho** [comment]
2.  I'd like to wrap some RR variables and methods into a new
"raconteur_handler" C++ class.  In particular, it'd be nice if this
new class could instantiate member var 


_xarm_rr_obj_ptr = RobotRaconteur::rr_cast<adi::xarm_rr::xarm_rr_obj>(
    RobotRaconteur::RobotRaconteurNode::s()->ConnectService(url));_

which could then be accessed by various methods that need to transmit
RR messages between my C++ client and Honglu's python server.

So I tried bundling up various setup calls within an initialization
member function:



_void raconteur_handler::initialize_member_objects()
{
   // Use node setup to help initialize client node (make file specific commands)
   service_types.push_back(RR_MAKE_SHARED<adi::xarm_rr::adi__xarm_rrFactory>());    

   // Use node setup to help initialize client node
//   RobotRaconteur::ClientNodeSetup node_setup(service_types);

   // Set time out for requests in milliseconds:
   RobotRaconteur::RobotRaconteurNode::s()->SetRequestTimeout(100000);

   // Connect to the Robot Raconteur service (make file specific commands)
   xarm_rr_obj_ptr =
      RobotRaconteur::rr_cast<adi::xarm_rr::xarm_rr_obj>(
         RobotRaconteur::RobotRaconteurNode::s()->ConnectService(url));
}_


But when xarm_rr_obj_ptr is used later, RR complains that some node
has gone out of scope.

I'm hoping there's some simple change(s) that can be made so that
xarm_rr_obj_ptr will persist for the life of the raconteur_handler
object.  Any suggestions?

**Reply by PeterCho** [reply]
Note:  There are a few extra underscore characters in my 2 questions above.  I was trying to italicize code snippets to make them a little easier to read.  But for the longer excertps, the underscores didn't turn into italicized text.

**Reply by johnwason** [reply]
I will have to look closely at this question.

I have been working on improved documentation will answer a lot of these questions when complete. Here is my progress: https://robotraconteur.github.io/robotraconteur/doc/core/latest/getting_started/

**Comment by PeterCho** [comment]
[raconteur_handler.h.txt](https://github.com/robotraconteur/robotraconteur/files/15188774/raconteur_handler.h.txt)
[raconteur_handler.cc.txt](https://github.com/robotraconteur/robotraconteur/files/15188780/raconteur_handler.cc.txt)

**Comment by PeterCho** [comment]
OK.  Honglu just figured out how to get around the linking problems we've
been seeing since yesterday.  Had to comment out a line in one of RR's
CmakeList.txt files.  But after that hack, we now have two way
communications between our robot arm which he's controlling via a python
script and my C++ app via RR working.  Yeah!  (And Honglu still has 1/2
left before he finishes his internship at Analog Garage...plenty of time
for us to make more progress...)

**Comment by PeterCho** [comment]
Honglu has 1/2 hour left before he finishes up in Boston...

**Comment by PeterCho** [comment]
This past week, Honglu and I had problems with RR linking errors on
machines operating under both ubuntu 22.4 and ubuntu 20.4.  On Friday,
Honglu said you had suggested replacing RR-companion installed via
apt-get with a manually installed version.  So we performed the
following procedure:


**sudo apt-get remove librobotraconteur-companion-dev
git clone https://github.com/robotraconteur/robotraconteur_companion.git --recursive
cd robotraconteur_companion/
cmake -S . -B build -DCMAKE_BUILD_TYPE=Debug
cmake --build build
cd build/
sudo make install**


After making this change, we were able to compile and link within our
cmake setup Honglu's C++ program that receives imagery from a camera
installed on a robot arm:

[image_view.cc.txt](https://github.com/robotraconteur/robotraconteur/files/15212688/image_view.cc.txt)


But when I try to build this same C++ app within my non-gitlab
development environment which works with a monolithic Makefile rather
than cmake, I'm seeing linking errors similar to those we encountered
last week on my ubuntu 20.4 machine:

![image_view_linking_errors](https://github.com/robotraconteur/robotraconteur/assets/3648253/7f7a61ff-ea6c-4917-9b4f-f9809a9288b3)

So I tried replacing robotraconteur-dev via on the 20.4 linux box via
a procedure analogous to the one that worked for
librobotraconteur-companion-dev:

**sudo apt-get remove robotraconteur-dev
git clone git@github.com:robotraconteur/robotraconteur.git
cd robotraconteur
cmake -S . -B build -DCMAKE_BUILD_TYPE=Debug
cmake --build build
cd build/
sudo make install**

Unfortunately, this did not resolve the linking problems.  Any
suggestions?

**Reply by johnwason** [reply]
I was having problems with the linking on the core RR library due to supporting both C++98 and >=C++11 linking formats. The fix required some changes, and it appears the companion library was not update. I have corrected the companion on the PPA, so it should be possible to update using APT and use the PPA version. Run `sudo make uninstall` to remove the locally built version and install `sudo apt install librobotraconteur-companion-dev`.

**Comment by PeterCho** [comment]
Here's the Makefile that was generated on my ubuntu 20.4 machine after I ran the first set of command lines that appear in bold font in my previous message (and which I got from Honglu on Friday):

[Makefile.txt](https://github.com/robotraconteur/robotraconteur/files/15214952/Makefile.txt)

It does not appear to contain any "uninstall" option.  So when I run "sudo make uninstall" from within robotraconteur_companion/build/, make returns the error *** No rule to make target 'uninstall'.    Is there some other easy way to get rid of all files that were recently installed by the first set of command lines in bolded font above?

**Reply by johnwason** [reply]
These commands should uninstall:

```
sudo rm -r /usr/local/lib/cmake/RobotRaconteurCompanion
sudo rm -r /usr/local/share/robotraconteur
sudo rm -r /usr/local/include/RobotRaconteurCompanion
sudo rm /usr/local/lib/libRobotRaconteurCompanion.a
```

**Comment by PeterCho** [comment]
Here are the steps I followed this morning:

1.  Remove "manually-installed" version of RobotRaconteurCompanion on
my ubuntu 20.4 machine using your commands above:

sudo rm -r /usr/local/lib/cmake/RobotRaconteurCompanion
sudo rm -r /usr/local/share/robotraconteur
sudo rm -r /usr/local/include/RobotRaconteurCompanion
sudo rm /usr/local/lib/libRobotRaconteurCompanion.a

2.  Reinstalled RobotRaconteurCompanion via ppa :

sudo add-apt-repository ppa:robotraconteur/ppa
sudo apt-get update
sudo apt-get install librobotraconteur-companion-dev

So version 0.3.1.6~focal of librobotraconteur-companion-dev is now
installed on my 20.4 box:

Westford:robot_raconteur% sudo apt show librobotraconteur-companion-dev
Package: librobotraconteur-companion-dev
Version: 0.3.1-6~focal

3.  Note: I'm still working with the "manually-installed" version of
robotraconteur-dev which I put onto the 20.4 machine via the command
lines from yesterday:

**sudo apt-get remove robotraconteur-dev
git clone git@github.com:robotraconteur/robotraconteur.git
cd robotraconteur
cmake -S . -B build -DCMAKE_BUILD_TYPE=Debug
cmake --build build
cd build/
sudo make install**

After making these changes, I am still able to rebuild Honglu's
"image_view.cc" program inside my machine's gitlab C++ environment
that works with Cmake.  But unfortunately, there are still lots of
linker errors when I try to build image_view.cc in my non-gitlab C++
setup which works with my monolithic Makefile:
![image
![image_view_02](https://github.com/robotraconteur/robotraconteur/assets/3648253/b4b4fd3a-46de-4608-8189-49795b9ab11e)
_view_01](https://github.com/robotraconteur/robotraconteur/assets/3648253/a1067e30-92ab-4929-89ca-08e251fce5af)

I do all C++ code and algorithm development in my non-gitlab / Makefile setup.  It's much less convoluted and simpler for me to understand than the gitlab / Cmake enviornment.  So I'd really prefer to be able to fully build image_view.cc in both setups if possible.

**Reply by johnwason** [reply]
I think you are missing the -DROBOTRACONTEUR_NO_CXX11 compiler flag. See https://github.com/robotraconteur/robotraconteur/wiki/Makefile-Compilation

I realized this flag wasn’t necessary for gcc 6.1 and greater so this will not be necessary after the next release of RR. The cmake will update automatically but you will need to remove it from your makefile.

**Reply by johnwason** [reply]
Oh sorry I didn’t read the log carefully enough. Gcc is very picky about the order of libraries, and I think you may need to list the companion library before core in the linker flags.

**Reply by PeterCho** [reply]
Yeah!.  Seems like that switch did the trick.  (If memory serves, libraries listed later are actually linked in earlier by gcc.  So makes sense that RR_core would need to come in before RR_companion.)  
Thanks for all the help.  Today I'm going to see if I can visualize video frames from an IntelRealSense camera that Honglu and Audren installed onto the robot arm a few weeks ago after flowing back to my C++ app via RR.  I won't use OpenCV.  But otherwise, Honglu basically showed this is possible last week.

**Comment by PeterCho** [comment]
Hi John. The flag you mention is in fact tacked on at the end of the CXXFLAGS var in my monolithic Makefile:
![cxx_flags_01](https://github.com/robotraconteur/robotraconteur/assets/3648253/ffcc4809-33eb-41ac-b170-06c88f0af84c)
![cxx_flags_02](https://github.com/robotraconteur/robotraconteur/assets/3648253/21ef4d79-74ae-42c5-be68-4c8438eddd19)

Since the problem I'm having is within linking rather than compiling, is CXXFLAGS a likely culprit for the source of the trouble with building RR on my ubuntu 20.4 machine outside of our lab's gitlab/cmake environment?

**Comment by johnwason** [comment]
I have released Robot Raconteur v1.2.1, which should fix the C++11 issues for GCC >=6.1. Ubuntu Bionic uses GCC 7.4 so all newer versions of Ubuntu should no longer need the `-DROBOTRACONTEUR_NO_CXX11` flag.
